### DISCUSSION TITLE
Create new record via API

### DISCUSSION THREAD

**Discussion by SaschaValentin** [pocketbase/pocketbase#2779] [Q&A] [answered]
Dear community,

really sorry have to ask such a question here but I feel like I'm completely screwed. I can´t get into my head how authentication over API works.

I have a collection "coll1" with an API rule `@request.auth.id != "" && @request.auth.role = "Gemeinde"` and created a user with this role. But how can this user now create new records? I´ve tried creating new entries over Postman without the API rule on the collection. This works. But as soon as I put the rule on and make a POST request with "username" and "password" in the body an error 400 "Failed to create record" appears.

I guess it´s because I put the credentials of the user in the false place. But where to enter them? Or is there something else wrong?

**Comment by ganigeorgiev** [comment]
> I can´t get into my head how authentication over API works.

A PocketBase request is considered authenticated if it has `Authorization: TOKEN` header.

The token can be generated using the admin or record auth apis. For more details you can check:

- https://pocketbase.io/docs/authentication/
- https://pocketbase.io/docs/api-records/#auth-record-actions
- https://pocketbase.io/docs/api-admins/

**Reply by SaschaValentin** [reply]
Thanks a lot for your quick help @ganigeorgiev .
I did this before with the admin but I didn´t thought it is the same with users. 

May I ask another similarly stupid question in this case using the Javascript SDK? I´m new to this and need just one litte kick to understand it.

When you have something like this

`import PocketBase from 'pocketbase';

const pb = new PocketBase('http://127.0.0.1:8090');

...

const authData = await pb.collection('users').authWithPassword(
    'YOUR_USERNAME_OR_EMAIL',
    'YOUR_PASSWORD',
);`

is this executed automatically? This sounds unprofessionally - let me explain: When I have a form where the user can enter his username and password then on button click this is send to Pocketbase and it returns the tolen?

**Reply by ganigeorgiev** [reply]
The above `authWithPassword()` call will send a `POST /api/collections/users/auth-with-password` request and return the response (token and record data) to the client.

Also note that when using the JS SDK, it comes with a builtin [AuthStore](https://github.com/pocketbase/js-sdk#authstore) that will keep the token and will send it automatically with every consequent request so you don't have to manually add the `Authorization` header.

**Reply by SaschaValentin** [reply]
Okay, I´ll try to get this runnig. Thanks a lot!

**Comment by SaschaValentin** [comment]
Dear @ganigeorgiev ,

the code examples in the API preview only show the use for the SDK in modules. I´m trying to rewrite

`const userData = await pb.collection('users').authWithPassword('test@example.com', '123456');`

to get it in an async function so I can use it without modules. I think it must be something like

`async function test(){
        const userData = await pb.collection('users').authWithPassword('username', 'password');
        let test = await userData;
    return test}
    test();`

But this also ends in a promise. But I´m trying to get the object so that I can access the token from it.

**Reply by ganigeorgiev** [reply, marked as answer]
I'm not sure I follow.

All SDK web API calls return a standard JavaScript [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).

You can either `await` the `Promise` if inside a `async` block (_or as top-level when using ESM_), or chain a `then` callback:

```js
pb.collection('users')
    .authWithPassword('test@example.com', '123456')
    .then((result) => {
        console.log(result)
    })
    .catch((err) => {
        console.log(err)
    })
```

If that's not what you are looking for, please elaborate and/or provide a more complete code sample what you are trying to do.
